### PR TITLE
Introduce jinterface.jl and small code refactoring

### PR DESCRIPTION
**Pull request by ysimillides**
Introduce jinterface.jl, which (will) provide ways to access FEniCS structures from Julia and incorporate them into our workflow. Currently we convert meshes to use them for FE equations.

### CODE REVIEW

File: src/jinterface.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+type feMesh
+    n_nodes::Int64
+    n_elements::Int64
+    nodes::Array{Float64, 2}
+    internal_nodes::Array{Int64}
+    boundary_nodes::Array{Int64}
+    elements::Array{Int64, 2}
+    n_internal_nodes::Int64
+    n_boundary_nodes::Int64
+    node_vals::Array{Float64}
```
Comment (ChrisRackauckas) [inline src/jinterface.jl]: `Vector`?

File: src/jinterface.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+Creates a type which incorporates the mesh attributes alongside the necessary values
+to solve finite element problems. It works by taking in a FEniCS mesh, working out
+the necessary values, and then creating an ordering of the internal/boundary nodes
+and applying the DirichletBC (via a function) to the respective nodes.
+"""
+type feMesh
+    n_nodes::Int64
+    n_elements::Int64
+    nodes::Array{Float64, 2}
+    internal_nodes::Array{Int64}
```
Comment (ChrisRackauckas) [inline src/jinterface.jl]: Note that arrays need dimension to be concretely typed, so accessing this is type-unstable.

File: src/jinterface.jl [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    boundary_nodes::Array{Int64}
+    elements::Array{Int64, 2}
```
Comment (ChrisRackauckas) [inline src/jinterface.jl]: For style, you can use the alias `Matrix`

File: src/jinterface.jl [hunk 4]
```diff
@@ -0,0 +1,69 @@
+"""
+Creates a type which incorporates the mesh attributes alongside the necessary values
+to solve finite element problems. It works by taking in a FEniCS mesh, working out
+the necessary values, and then creating an ordering of the internal/boundary nodes
+and applying the DirichletBC (via a function) to the respective nodes.
+"""
+type feMesh
+    n_nodes::Int64
```
Comment (ChrisRackauckas) [inline src/jinterface.jl]: Is 64 for C-operability? Then you might want to use `CInt`. If on 32-bit systems it should use 32-bit integers (which is possibly the case?), then use `Int`. If FEniCS doesn't work on 32-bit, then none of this matters but for style I would use `Int`.

File: src/jinterface.jl [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    n_elements::Int64
+    nodes::Array{Float64, 2}
+    internal_nodes::Array{Int64}
+    boundary_nodes::Array{Int64}
+    elements::Array{Int64, 2}
+    n_internal_nodes::Int64
+    n_boundary_nodes::Int64
+    node_vals::Array{Float64}
+
+    function feMesh(mesh::Mesh,boundaryCondition::Core.Function)
```
Comment (ChrisRackauckas) [inline src/jinterface.jl]: `Base.Function`?

File: src/jinterface.jl (hunk 4 above)
Comment (ysimillides) [reply src/jinterface.jl]: Will change to Int to allow operability on both.

File: src/jinterface.jl (hunk 5 above)
Comment (ysimillides) [reply src/jinterface.jl]: Corrected ( will also check for any other Core / Base discrepancies in the code base

File: src/jinterface.jl (hunk 1 above)
Comment (ysimillides) [reply src/jinterface.jl]: Changed the Array{Float64} to Vector

File: src/jinterface.jl (hunk 2 above)
Comment (ysimillides) [reply src/jinterface.jl]: Something like nodes::Array{Float64, 2::Int} or am I mistaken? Also for the Alias would it then be nodes::Matrix{Float64}? or preferably nodes::Matrix{AbstractFloat}?

Comment (ChrisRackauckas) [reply src/jinterface.jl]: Just `Array{Float64,2}` or `Matrix{Float64}`. No assertions in there since that `2` is a value and not a type location.

**Pull request merged**